src/manifolds/ProbabilitySimplex.jl

```diff
@@ -0,0 +1,294 @@
+        Tuple{typeof(get_embedding(M)), typeof(p), typeof(X)},
+        get_embedding(M),
+        p,
+        X;
+        check_base_point = false, # already checked above
+        kwargs...
+    )
+    mpv === nothing || return mpv
+    if !isapprox(sum(X), 0.0; kwargs...)
+        return DomainError(
+            sum(X),
+            "The vector $(X) is not a tangent vector to $(p) on $(M), since its elements to not sum up to 0.",
+        )
+    end
+    return nothing
+end
+
+decorated_manifold(M::ProbabilitySimplex) = Euclidean(representation_size(M)...; field = ℝ)
+
+default_metric_dispatch(::ProbabilitySimplex, ::FisherRaoMetric) = Val(true)
+
+@doc raw"""
+    distance(M,p,q)
+
+Compute the distance between two points on the [`ProbabilitySimplex`](@ref) `M`.
+The formula reads
+````math
+d_{Δ^n}(p,q) = 2\arccos \biggl( \sum_{i=1}^{n+1} \sqrt{p_i q_i} \biggr)
+````
+"""
+function distance(::ProbabilitySimplex, p, q)
+    sumsqrt = zero(Base.promote_eltype(p, q))
+    @inbounds for i in eachindex(p, q)
+        sumsqrt += sqrt(p[i] * q[i])
+    end
+    return 2*acos(sumsqrt)
+end
+
+embed!(::ProbabilitySimplex, q, p) = copyto!(q, p)
+embed!(::ProbabilitySimplex, Y, p, X) = copyto!(Y, X)
+
+@doc raw"""
+    exp(M::ProbabilitySimplex,p,X)
+
+Compute the exponential map on the probability simplex.
+
+````math
+\exp_pX = \frac{1}{2}\Bigl(p+\frac{X_p^2}{\lVert X_p \rVert^2}\Bigr)
++ \frac{1}{2}\Bigl(p - \frac{X_p^2}{\lVert X_p \rVert^2}\Bigr)\cos(\lVert X_p\rVert)
++ \frac{1}{\lVert X_p \rVert}\sqrt{p}\sin(\lVert X_p\rVert),
+````
+
+where $X_p = \frac{X}{\sqrt{p}}$, with its division meant elementwise, as well as for the
+operations $X_p^2$ and $\sqrt{p}$.
+"""
+exp(::ProbabilitySimplex, ::Any...)
+
+function exp!(::ProbabilitySimplex, q, p, X)
+    s = sqrt.(p)
+    Xs = X ./ s ./ 2
+    θ = norm(Xs)
+    q .= (cos(θ) .* s .+ usinc(θ) .* Xs).^2
+    return q
+end
+
+@doc raw"""
+    inner(M::ProbabilitySimplex,p,X,Y)
+
+Compute the inner product of two tangent vectors `X`, `Y` from the tangent space $T_pΔ^n$ at
+`p`. The formula reads
+````math
+g_p(X,Y) = \sum_{i=1}^{n+1}\frac{X_iY_i}{p}
+````
+"""
+function inner(::ProbabilitySimplex, p, X, Y)
+    d = zero(Base.promote_eltype(p,X,Y))
+    @inbounds for i in eachindex(p,X,Y)
+        d += X[i]*Y[i]/p[i]
+    end
+    return d
+end
+
+@doc raw"""
+    inverse_retract(M::ProbabilitySimplex, p, q, ::SoftmaxInverseRetraction)
+
+Compute a first order approximation by projection. The formula reads
+````math
+\operatorname{retr}^{-1}_p q = \bigl( I - \frac{1}{n}\mathbb{1}\mathbb{1}^\mathrm{T}\bigr)(\log(q)-\log(p))
+````
+where $\mathbb{1}$ is the column vector containing ones and $\log$ is applied elementwise.
+"""
+inverse_retract(::ProbabilitySimplex, ::Any, ::Any, ::SoftmaxInverseRetraction)
+
+function inverse_retract!(::ProbabilitySimplex{n}, X, p, q, ::SoftmaxInverseRetraction) where {n}
+    X .= (one(zeros(n+1,n+1)) - 1/n.*ones(n+1,n+1) ) * (log.(q) - log.(p))
```

Comment:
```suggestion
    logdiff = log.(q) .- log.(p)
    meanlogdiff = mean(logdiff)
    copyto!(X, logdiff)
    X .-= meanlogdiff
```
This looks like it gives the same result but is much faster.

Reply:
Oh, it can actually be even faster.

Reply:
Cool! Thanks for checking; Yes, my version built whole matrices, that was fore sure not efficient.